src/_P164_gases_ens160.ino

```diff
@@ -0,0 +1,158 @@
+
+# include "src/PluginStructs/P164_data_struct.h"
+
+# define PLUGIN_164
+# define PLUGIN_ID_164         164
+# define PLUGIN_NAME_164       "Gases - ENS16x"
+# define PLUGIN_VALUENAME1_164 "TVOC"
+# define PLUGIN_VALUENAME2_164 "eCO2"
+
+boolean Plugin_164(uint8_t function, struct EventStruct *event, String& string)
+{
+  boolean success = false;
+
+  switch (function)
+  {
+    case PLUGIN_DEVICE_ADD:
+    {
+      Device[++deviceCount].Number           = PLUGIN_ID_164;
+      Device[deviceCount].Type               = DEVICE_TYPE_I2C;
+      Device[deviceCount].VType              = Sensor_VType::SENSOR_TYPE_DUAL;
+      Device[deviceCount].Ports              = 0;
+      Device[deviceCount].PullUpOption       = false;
+      Device[deviceCount].InverseLogicOption = false;
+      Device[deviceCount].FormulaOption      = true;
+      Device[deviceCount].ValueCount         = 2;
+      Device[deviceCount].SendDataOption     = true;
+      Device[deviceCount].TimerOption        = true;
+      Device[deviceCount].GlobalSyncOption   = true;
+      Device[deviceCount].PluginStats        = true;
+      Device[deviceCount].I2CNoDeviceCheck   = true;  //TODO
+      break;
+    }
+
+    case PLUGIN_GET_DEVICENAME:
+    {
+      string = F(PLUGIN_NAME_164);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+    {
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_164));
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_164));
+      break;
+    }
+
+    case PLUGIN_I2C_HAS_ADDRESS:
+    case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+    {
+      const uint8_t i2cAddressValues[] = { ENS160_I2CADDR_0, ENS160_I2CADDR_1 };
+      constexpr int nrAddressOptions   = NR_ELEMENTS(i2cAddressValues);
+
+      if (function == PLUGIN_WEBFORM_SHOW_I2C_PARAMS) {
+        addFormSelectorI2C(F("i2c_addr"), nrAddressOptions, i2cAddressValues, P164_I2C_ADDR);
+        addFormNote(F("ADDR Low=0x52, High=0x53"));
+      } else {
+        success = intArrayContains(nrAddressOptions, i2cAddressValues, event->Par1);
+      }
+
+      break;
+    }
+
+    # if FEATURE_I2C_GET_ADDRESS
+    case PLUGIN_I2C_GET_ADDRESS:
+    {
+      event->Par1 = P164_I2C_ADDR;
+      success     = true;
+      break;
+    }
+    # endif // if FEATURE_I2C_GET_ADDRESS
+
+    case PLUGIN_SET_DEFAULTS:
+    {
+      P164_I2C_ADDR = ENS160_I2CADDR_1;
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      initPluginTaskData(event->TaskIndex, new (std::nothrow) P164_data_struct(event));
+      P164_data_struct *P164_data = static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      success = (nullptr != P164_data && P164_data->begin());
+      break;
+    }
+
+    case PLUGIN_READ:
+    {
+      P164_data_struct *P164_data = static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr == P164_data) {
+        Serial.print("P164: plugin_read NULLPTR");
```

Comment:
Please don't output to Serial.
Rather use `addLog(LOG_LEVEL_ERROR, F("...."));`
`Serial` could be in use for a sensor and sending logs to it will disrupt operations with such a sensor.